### PR TITLE
Make text formatting bar sticky, and introduce query mixin

### PR DESCRIPTION
**Pull request by jasmussen**
This PR adds a bunch of media query variables and a mixin for using them, and rewrites current queries to use this new method.

It also makes the dummy toolbar for HTML editing sticky.

### CODE REVIEW

File: editor/modes/text-editor/index.js [hunk 1]
```diff
@@ -13,19 +13,21 @@ function TextEditor( { blocks, onChange } ) {
...
-				<button className="editor-text-editor__del">del</button>
-				<button>ins</button>
-				<button>img</button>
-				<button>ul</button>
-				<button>ol</button>
-				<button>li</button>
-				<button>code</button>
-				<button>more</button>
-				<button>close tags</button>
+				<div className="editor-text-editor__formatting-inner">
```
Comment (aduth) [inline editor/modes/text-editor/index.js]: In my effort to rid this world of superfluous wrapping components, I might suggest the following alternatives to the issue of padding being affected by the left offset:

- Move the entire editor canvas into a fixed position container and assign the headers as top-assigned absolute within. It's probably not great for maintainability to have non-root components consider sidebar and admin bar offsets.
- Use a different unit for the padding: a pixel or `vw` value perhaps

File: webpack.config.js [hunk 2]
```diff
@@ -45,7 +45,7 @@ const config = {
 							loader: 'sass-loader',
 							query: {
 								includePaths: [ 'editor/assets/stylesheets' ],
-								data: '@import "variables";',
+								data: '@import "variables"; @import "_mixins";',
```
Comment (aduth) [inline webpack.config.js]: Two things:

- IIRC, SASS doesn't care whether the file is prefixed with an `_` and will allow `@import "mixins";` here?
- We should choose one or the other for the file name convention and apply it to `variables`

File: editor/modes/text-editor/index.js (hunk 1 above)
Comment (jasmussen) [inline editor/modes/text-editor/index.js]: > Move the entire editor canvas into a fixed position container and assign the headers as top-assigned absolute within. It's probably not great for maintainability to have non-root components consider sidebar and admin bar offsets.

This is interesting. I'm a little nervous about using too much fixed position, though, since css transitions (as I've learned and I'm sure you recall because you graciously fixed it for me) affect fixed position. Might there be sideeffects like that here, should we use CSS transitions on content inside the canvas?

Also, in the future we're likely to have more fixed or sticky position stuff like this. For example if you have a very very long text block where you need to scroll, as you scroll down the page, the Quick Toolbar (<a href="https://wpcoredesign.mystagingwebsite.com/gutenberg/#blueprints">new name for the controls docked to the top of the block</a>) would unstick and dock themselves to the top of the screen instead. Like the Calypso editor, in a way. 

I'm not objecting, just trying to wrap my head around all the components. 

> Use a different unit for the padding: a pixel or vw value perhaps

The problem isn't so much the "faux-centering" the padding does of the contents within, it's rather that on very large breakpoints I mean to apply a max-width to the text inside the editor:

screen shot 2017-04-14 at 10 15 10

I should have clarified this in the past, sorry. But this is the code in question:

```
.editor-text-editor__formatting,
.editor-text-editor {
	padding-left: $item-spacing * 2;
	padding-right: $item-spacing * 2;
	max-width: $break-huge;
	margin: 0 auto;

	@include break-large() {
		padding-left: 10%;
		padding-right: 10%;
	}

	@include break-wide() {
		padding-left: 15%;
		padding-right: 15%;
	}

	@include break-huge() {
		padding-left: 200px;
		padding-right: 200px;
	}
}
```

As you can see in the `break-huge` breakpoint (1440px), the padding becomes a px unit. Combined with a max-width also of 1440 and `margin: 0 auto;`, that centers the text editor and doesn't let it grow larger. This is working fine, but as you can see on the docked toolbar, the gray border below it doesn't extend fully left and right. I suppose I could potentially do some `:before` magic to do that, instead of having a superfluous container. But that, specifically, is the problem I'm trying to solve.

File: webpack.config.js (hunk 2 above)
Comment (jasmussen) [inline webpack.config.js]: Addressed this by moving to underscores.

**Comment by jasmussen** [comment]
Great feedback, than you for the review. Addressed the webpack stuff.

I'd like some thoughts on how best to proceed with the problem outlined in https://github.com/WordPress/gutenberg/pull/417#discussion_r111545141. The easy fix is to keep the superfluous div intact, but I understand the concern and perhaps we can think up an architectural change that might allow us to do the same in a better way. Thoughts welcome.

**Comment by aduth** [comment]
@jasmussen Am I correct in assuming we only need the fixed position in the first place to ensure that the entire background is white? If so, could we instead override the background color of the `body` directly in the context of the Gutenberg page, and avoid the need for any fixed elements?

File: webpack.config.js [hunk 3]
```diff
@@ -45,7 +45,7 @@ const config = {
 							loader: 'sass-loader',
 							query: {
 								includePaths: [ 'editor/assets/stylesheets' ],
-								data: '@import "variables";',
+								data: '@import "variables"; @import "mixins";',
```
Comment (aduth) [inline webpack.config.js]: We still have inconsistency between usage of underscore on the `variables.scss` and `_mixins.scss` files themselves.

**Comment by aduth** [comment]
Aside from naming consistency, I think this is fine as is. The only other idea I'd have with regards to fixed positioning is we could create a separate scrollbar context for the editor canvas and leave the formatting toolbar as static positioned. The fixed stuff is complex to maintain. I think we'll also have to revisit this once we start implementing the post settings panel.

**Comment by aduth** [comment]
Fake fixed header concept with separate scroll context: https://codepen.io/aduth/pen/GmJbvR

File: webpack.config.js (hunk 3 above)
Comment (jasmussen) [inline webpack.config.js]: Apologies for being daft here, but I can't find what you're referring to. It should be `_variables.scss` and `_mixins.scss` and `main.scss`, correct? 

screen shot 2017-04-18 at 09 41 31

**Comment by jasmussen** [comment]
Rebased, which was a bad idea, because now it picks up a change in master which broke the text editor mode. Silly because I think I have an idea for how to remove the superfluous `inner` div, while still having a bottom border that extends to the edges on >1440px breakpoints. Basically using :before and :after flexed spacers that happen to be the same width as the textarea padding. 

I share your frustration with `position: fixed;` — the fact that it can't have positioning values relative to its parent container makes for ballooningly complex CSS. Using `overflow: auto;` in a container does certainly fix that, however sadly this does not work well with iOS, even if we apply `-webkit-overflow-scrolling: touch;` to the scrollable area. 

The problem is demonstrated in this video, which shows my attempt at doing just that for Calypso:

https://videopress.com/v/c7xHUdNc

In comparison, here's how it should work:

https://videopress.com/v/oW7pev5L

The problem is the overflow scroll bounce that iOS has —  the one where if you touch drag near at the start or at the end of a scroll context, the entire canvas inside is "elastically pulled down" to indicate there is no more content to scroll —  that effect triggers even on contexts that do not need scrolling, i.e. the surrounding page itself. 

So say you were using the editor on an iPad, and meant to scroll to the bottom. Let's say you accidentally touch-dragged the left navigation menu instead of the editable field, what would happen is you'd try and scroll the `html` context, which doesn't need scrolling, and you'd bounce things around.

And so even if your codepen example would work great on a desktop breakpoint, we'd still need a hack for mobile. Sadly, for now, the reality is that for the experience to be good on iOS we probably want to scroll the `html` context, and use `position: fixed;` for any elements that need to stay in place.

**Comment by jasmussen** [comment]
I pushed a CSS trick that centers the toolbar contents using invisible spacers. This allows us to rid of the "inner" element. Let me know your thoughts.

File: webpack.config.js (hunk 3 above)
Comment (aduth) [inline webpack.config.js]: Hmm, I swear when I'd last looked one had the underscore and the other not, but it appears correct now, yes.

File: editor/modes/text-editor/style.scss [hunk 4]
```diff
@@ -59,40 +87,73 @@
...
+		&:before,
+		&:after {
+			flex: 0 1 10%;
+		}
+	}
+
+	@include break-wide() {
+		&:before,
+		&:after {
+			flex: 0 1 15%;
```
Comment (aduth) [inline editor/modes/text-editor/style.scss]: What is the reason this needs to be flex vs. simply `width: 15%;` ? Does the `flex-shrink` play into this?

**Review by aduth — APPROVED** [review]
A question about needing Flex, but otherwise this LGTM 👍  And thanks for the detail about `overflow` on iOS.

File: editor/modes/text-editor/style.scss (hunk 4 above)
Comment (jasmussen) [inline editor/modes/text-editor/style.scss]: Just realized that a) you're right, width could have worked just as well here, and b) this doesn't actually solve the problem I was trying to solve. I'm going to keep working for a bit here. I must have lacked more sleep than I thought I lacked. Thanks for the reviews and apologies, going to bake this one a bit more.

**Comment by jasmussen** [comment]
What a roller-coaster this branch has been. That'll learn me to make two PRs in the future, one for the mixin, one for the fixed position header.

The latest commit fixes the problem, without having a superfluous div. But it also perfectly demonstrates your point, @aduth, that fixed positioning due to its lack of relativism to parent widths, is complex. 

Cases in point, https://github.com/WordPress/gutenberg/pull/417/commits/ac2af342e8e9918a140a0959bd80bae95ceaffec#diff-fae2b472a4c5ff66153906e36fd0be71R113, and https://github.com/WordPress/gutenberg/pull/417/commits/ac2af342e8e9918a140a0959bd80bae95ceaffec#diff-fae2b472a4c5ff66153906e36fd0be71R120. Here are the pertinent bits:

```
.editor-text-editor {
	@include break-large() {
		padding-left: calc( 50% - 400px + 20px );		// 50% minus half of the max-width of the main column
		padding-right:  calc( 50% - 400px + 20px );
	}
}

.editor-text-editor__formatting {
	@include break-large() {
		padding-left: calc( 50% - 400px + 20px - 80px );		// 50% minus half of the max-width of the main column, minus half of the left offset
		padding-right:  calc( 50% - 400px + 20px - 80px );
	}
}
```

In the first blob, we have `.editor-text-editor`. That's the textarea. We apply a padding to this so that no matter where you click in the textarea, you set focus on it. 

By setting the left and right paddings to this calc width, we effectively mimic a max-width. `calc( 50% - 400px + 20px )` is 50% of the element width, minus half of the desired max-width, + 20px as a baseline minimal margin. 

To explain it more simply, it behaves like a `max-width: 800px` does, same as `width: calc( 100% - 800px )` would do. Just divided in two for left and right paddings. Hacky, yes. But it works. 

The problem is that `.editor-text-editor__formatting` is fixed position. So 50% for that element is more than 50% of the textarea (because it calculates 50% off of HTML instead of `.gutenberg`), thus beckoning yet another specific rule that takes the discrepancy into account. 

In short: fixed positioning math sucks.

```
............................................________........................ 
....................................,.-‘”...................``~.,.................. 
.............................,.-”...................................“-.,............ 
.........................,/...............................................”:,........ 
.....................,?......................................................\,..... 
.................../...........................................................,}.... 
................./......................................................,:`^`..}.... 
.............../...................................................,:”........./..... 
..............?.....__.........................................:`.........../..... 
............./__.(.....“~-,_..............................,:`........../........ 
.........../(_....”~,_........“~,_....................,:`........_/........... 
..........{.._$;_......”=,_.......“-,_.......,.-~-,},.~”;/....}........... 
...........((.....*~_.......”=-._......“;,,./`..../”............../............ 
...,,,___.\`~,......“~.,....................`.....}............../............. 
............(....`=-,,.......`........................(......;_,,-”............... 
............/.`~,......`-...............................\....../\................... 
.............\`~.*-,.....................................|,./.....\,__........... 
,,_..........}.>-._\...................................|..............`=~-,.... 
.....`=~-,_\_......`\,.................................\........................ 
...................`=~-,,.\,...............................\....................... 
................................`:,,...........................`\..............__.. 
.....................................`=-,...................,%`>--==``....... 
........................................_\..........._,-%.......`\............... 
...................................,<`.._|_,-&``................`\..............
```

**Where do we go from here?**

Keeping <a href="https://github.com/WordPress/gutenberg/pull/417#issuecomment-294720031">the aforementioned iOS issues</a> in mind, there are a number of ways we can go about this:

1. We can decide to not have fixed position on iOS, which would be a bummer as that's where it feels the most needed. It also doesn't solve the iPad problem, which is bigger than mobile.

2. We embrace the complexities of position: fixed, and try to simplify the math using sass variables or mixins or something, try to reduce the use as much as possible. But even then, this is going to get increasingly complex as we introduce a right-hand sidebar. 

3. ?? Magic ?? —  am I missing something here? The biggest problem with position: fixed is the **width** of the fixed positioned element, as we have to take into account a left hand navigation sidebar that varies in width depending on viewports. 

As a first step, and to get this merged, I'd like to keep the calc-padding hack for the textarea, because it works well, but re-introduce the superfluous div for the HTML toolbar (as this seems to go to too great lengths just to remove an extraneous element).

**Review by youknowriad — APPROVED** [review]
This behaves great. 

Yes, we should avoid unnecessary dom elements but I'd go for simpler CSS rather than harder to understand and maintain one. Let's ship this. And as you said, we can revisit anytime.

**Comment by jasmussen** [comment]
Thanks Riad!
